### PR TITLE
Proportional scaling in Deployments

### PR DESCRIPTION
**Pull request by kargakis**
@kubernetes/deployment ptal

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1653)

### CODE REVIEW

**Comment by devin-donnelly** [comment]
LGTM from a docs perspective. Could somebody who knows more about it than me give it a once-over?

**Comment by jaredbhatti** [comment]
@janetkuo Can you give a tech review?

**Comment by kargakis** [comment]
@mfojtik can you have a tech review here?

File: docs/user-guide/deployments.md [hunk 1]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
 
 Note: You cannot rollback a paused Deployment until you resume it.
 
+## Proportional scaling
+
+Rolling deployments support running mulitple versions of an application at the same time. When there is a
```
Comment (mfojtik) [inline docs/user-guide/deployments.md]: multiple

File: docs/user-guide/deployments.md [hunk 2]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
...
+new rollout ongoing or it is paused (you have multiple running versions in both cases) and a scaling request
+is received (either user-initiated or by an autoscaler), then the deployment controller will balance the
+additional replicas in the existing active replica sets in order to mitigate risk.
+
+For example, you are running a Deployment with 10 replicas, maxSurge=3, maxUnavailable=2.
```
Comment (mfojtik) [inline docs/user-guide/deployments.md]: i would put extra sentence to explain maxSurge and maxUnavailable

File: docs/user-guide/deployments.md [hunk 3]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
...
+nginx     10        10        10           10          50s
+```
+
+You update to a new image which happens to be unresolvable from inside the cluster.
+```
+$ kubectl set image deploy/nginx nginx=nginx:sometag
+deployment "nginx" image updated
+```
+
+The image update starts a new rollout but it's blocked due to the maxUnavailable requirement that we
```
Comment (mfojtik) [inline docs/user-guide/deployments.md]: point to the RS version that represents new rollout

File: docs/user-guide/deployments.md [hunk 4]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
...
+mentioned above.
+```
+$ kubectl get rs
+NAME               DESIRED   CURRENT   READY     AGE
+nginx-1989198191   5         5         0         9s
+nginx-618515232    8         8         8         1m
+```
+
+Then a new scaling request comes along. An autoscaler increments the Deployment replicas to 15. The
```
Comment (mfojtik) [inline docs/user-guide/deployments.md]: scaling request for deployment

**Comment by mfojtik** [comment]
few nits, lgtm otherwise :)

**Comment by kargakis** [comment]
Comments addressed.

**Comment by kargakis** [comment]
@jaredbhatti can we get this merged?

**Comment by kargakis** [comment]
@janetkuo can we get this merged

File: docs/user-guide/deployments.md [hunk 5]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
...
 
+## Proportional scaling
+
+RollingUpdate Deployments support running multitple versions of an application at the same time. When there
+is a new rollout ongoing or it is paused (you have multiple running versions in both cases) and a scaling
+request is received (either user-initiated or by an autoscaler), then the deployment controller will balance
+the additional replicas in the existing active ReplicaSets (ReplicaSets with Pods) in order to mitigate risk.
+
+For example, you are running a Deployment with 10 replicas, [maxSurge](#max-surge)=3, [maxUnavailable](#max-unavailable)=2.
+```
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: You need an additional line before ``` to make it render well (and other places in this change)

File: docs/user-guide/deployments.md [hunk 6]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
 
 Note: You cannot rollback a paused Deployment until you resume it.
 
+## Proportional scaling
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: `Scaling a Deployment`

File: docs/user-guide/deployments.md [hunk 7]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
...
+
+RollingUpdate Deployments support running multitple versions of an application at the same time. When there
+is a new rollout ongoing or it is paused (you have multiple running versions in both cases) and a scaling
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Suggest make it shorter and easier to read, something like "When you scale a Deployment that is either still in progress or paused, ..."

File: docs/user-guide/deployments.md (hunk 6 above)
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Talk about scale in general, and then talk about proportional scaling

File: docs/user-guide/deployments.md [hunk 8]
```diff
@@ -453,6 +453,55 @@ nginx-deployment-3066724191   0         0         1h
...
+request is received (either user-initiated or by an autoscaler), then the deployment controller will balance
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: Deployment controller

**Comment by kargakis** [comment]
@janetkuo thanks for your comments, updated

File: docs/user-guide/deployments.md [hunk 9]
```diff
@@ -390,6 +390,76 @@ Events:
...
+NAME                 DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
+nginx-deployment     15        18        7            8           7m
+$ kubectl get rs
+NAME                          DESIRED   CURRENT   READY     AGE
+nginx-deployment-1989198191   7         7         0         7m
+nginx-deployment-618515232    11        11        11        7m
+```
+
+
 ### Clean up Policy
```
Comment (janetkuo) [inline docs/user-guide/deployments.md]: "Clean up policy" belongs to the previous section. Please move the scaling section below it.

**Comment by janetkuo** [comment]
One last comment, LGTM otherwise

File: docs/user-guide/deployments.md (hunk 9 above)
Comment (kargakis) [inline docs/user-guide/deployments.md]: oops, fixing shortly

**Comment by kargakis** [comment]
@janetkuo done

**Comment by kargakis** [comment]
@david-mcmahon can we get this merged? Thanks

**Comment by kargakis** [comment]
> @david-mcmahon can we get this merged? Thanks

oops, meant @devin-donnelly

**Pull request merged**